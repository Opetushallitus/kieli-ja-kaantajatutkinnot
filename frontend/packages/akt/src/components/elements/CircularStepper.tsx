import { CircularProgress, CircularProgressProps } from '@mui/material';

import { H3 } from 'components/elements/Text';
import { useCommonTranslation } from 'configs/i18n';
import { Color } from 'enums/app';

interface CircularStepperProps extends CircularProgressProps {
  value: number;
  phaseText: string;
  phaseDescription: string;
}

export const CircularStepper = ({
  phaseText,
  phaseDescription,
  ...rest
}: CircularStepperProps) => {
  const translateCommon = useCommonTranslation();
  const ariaLabel = `${translateCommon(
    'phase'
  )} ${phaseText}: ${phaseDescription}`;

  return (
    <div className="circular-stepper">
      <CircularProgress
        aria-label={ariaLabel}
        className="circular-stepper__progress"
        color={Color.Secondary}
        variant="determinate"
        {...rest}
      />
      <div className="circular-stepper__heading">
        <H3>{phaseText}</H3>
      </div>
    </div>
  );
};
