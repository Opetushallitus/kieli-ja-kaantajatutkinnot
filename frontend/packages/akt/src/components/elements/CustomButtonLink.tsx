import { Button, ButtonProps } from '@mui/material';
import { FC } from 'react';
import { Link, LinkProps } from 'react-router-dom';

type CustomButtonLinkProps = ButtonProps &
  Pick<LinkProps, 'target' | 'rel' | 'to'>;

export const CustomButtonLink: FC<CustomButtonLinkProps> = ({
  to,
  target,
  rel,
  ...btnProps
}) => {
  return (
    <Link className="custom-button-link" to={to} target={target} rel={rel}>
      <Button {...btnProps}></Button>
    </Link>
  );
};
