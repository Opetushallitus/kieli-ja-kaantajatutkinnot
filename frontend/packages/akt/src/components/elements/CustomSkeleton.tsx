import { Skeleton, SkeletonProps } from '@mui/material';
import { FC } from 'react';

import { useCommonTranslation } from 'configs/i18n';

export const CustomSkeleton: FC<SkeletonProps> = ({ ...props }) => {
  const translateCommon = useCommonTranslation();

  return <Skeleton {...props} aria-label={translateCommon('loadingContent')} />;
};
