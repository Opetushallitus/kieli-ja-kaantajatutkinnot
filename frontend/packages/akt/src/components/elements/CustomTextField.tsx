import { TextField } from '@mui/material';

import { CustomTextFieldProps } from 'interfaces/components/customTextField';

export const CustomTextField = ({
  error,
  helperText,
  showHelperText,
  multiline,
  ...rest
}: CustomTextFieldProps) => {
  const minRows = multiline ? 5 : undefined;
  const maxRows = multiline ? 15 : undefined;

  return (
    <TextField
      minRows={minRows}
      maxRows={maxRows}
      error={error}
      multiline={multiline}
      helperText={(error || showHelperText) && helperText}
      {...rest}
    />
  );
};
