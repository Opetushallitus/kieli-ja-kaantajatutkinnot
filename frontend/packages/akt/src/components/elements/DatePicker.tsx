import Stack from '@mui/material/Stack';
import TextField from '@mui/material/TextField';
import { Dayjs } from 'dayjs';

import { DateUtils } from 'utils/date';

export interface DatePickerProps {
  value?: string;
  setValue: (value: string) => void;
  label: string;
  disabled?: boolean;
  minDate?: Dayjs;
  maxDate?: Dayjs;
  dataTestId?: string;
}

export const DatePicker = ({
  value,
  setValue,
  label,
  disabled = false,
  minDate,
  maxDate,
  dataTestId,
}: DatePickerProps): JSX.Element => {
  const MIN_DATE = '1900-01-01';
  const MAX_DATE = '2100-01-01';
  const dayjs = DateUtils.dayjs();

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
  };

  return (
    <Stack spacing={3}>
      <TextField
        data-testid={dataTestId ? `${dataTestId}__date-picker` : 'date-picker'}
        disabled={disabled}
        label={label}
        type="date"
        onChange={handleChange}
        value={value}
        InputLabelProps={{
          shrink: true,
        }}
        inputProps={{
          min: DateUtils.formatOptionalDate(minDate ?? dayjs(MIN_DATE)),
          max: DateUtils.formatOptionalDate(maxDate ?? dayjs(MAX_DATE)),
        }}
      />
    </Stack>
  );
};
