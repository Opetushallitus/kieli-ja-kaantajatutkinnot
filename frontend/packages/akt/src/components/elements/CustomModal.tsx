import Modal, { ModalProps } from '@mui/material/Modal';
import { FC } from 'react';

import { H2 } from 'components/elements/Text';

type CustomModalProps = ModalProps & {
  ariaLabelledBy?: string;
  screenReaderTitle?: string;
  modalTitle?: string;
  onCloseModal: () => void;
};

export const CustomModal: FC<CustomModalProps> = ({
  ariaLabelledBy,
  open,
  children,
  modalTitle,
  onCloseModal,
}) => {
  const handleOnClose = (event: React.SyntheticEvent, reason: string) => {
    if (event && reason === 'backdropClick') {
      return;
    }
    onCloseModal();
  };

  return (
    <Modal open={open} onClose={handleOnClose} aria-labelledby={ariaLabelledBy}>
      <div className="custom-modal">
        <div className="rows gapped">
          <H2 id={ariaLabelledBy}>{modalTitle}</H2>
          {children}
        </div>
      </div>
    </Modal>
  );
};
