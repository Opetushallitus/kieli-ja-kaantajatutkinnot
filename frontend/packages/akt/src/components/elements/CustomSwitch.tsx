import {
  FormControlLabel,
  FormGroup,
  Switch,
  SwitchProps,
} from '@mui/material';

import { Caption, Text } from 'components/elements/Text';
import { Color } from 'enums/app';

interface CustomSwitchProps extends SwitchProps {
  dataTestId?: string;
  leftLabel: string;
  rightLabel: string;
  errorLabel?: string;
  value?: boolean;
}

export const CustomSwitch = ({
  dataTestId,
  leftLabel,
  rightLabel,
  errorLabel,
  value,
  disabled,
  onChange,
}: CustomSwitchProps) => {
  const leftLabelClassName = disabled
    ? 'color-disabled margin-right-xs'
    : 'margin-right-xs';

  return (
    <FormGroup className="rows">
      <div className="columns">
        <Text className={leftLabelClassName}>{leftLabel}</Text>
        <FormControlLabel
          disabled={disabled}
          control={
            <Switch
              data-testid={dataTestId}
              checked={value}
              color={Color.Secondary}
              onChange={onChange}
            />
          }
          label={rightLabel}
        />
      </div>
      {errorLabel && (
        <Caption data-testid={`${dataTestId}__error-label`} color={Color.Error}>
          {errorLabel}
        </Caption>
      )}
    </FormGroup>
  );
};
