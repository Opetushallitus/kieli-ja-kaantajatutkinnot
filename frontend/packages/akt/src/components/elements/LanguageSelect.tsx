import { ComboBox, sortOptionsByLabels } from 'components/elements/ComboBox';
import { useKoodistoLanguagesTranslation } from 'configs/i18n';
import {
  AutoCompleteComboBox,
  ComboBoxOption,
  ComboBoxProps,
} from 'interfaces/components/combobox';

export interface LanguageSelectProps {
  excludedLanguage: string | undefined;
  languages: Array<string>;
}

const primaryLanguages = ['FI', 'SV', 'SEIN', 'SEKO', 'SEPO'];

export const languageToComboBoxOption = (
  translate: (l: string) => string,
  lang: string
): ComboBoxOption => ({
  label: translate(lang),
  value: lang,
});

export const LanguageSelect = ({
  excludedLanguage,
  languages,
  ...rest
}: LanguageSelectProps &
  Omit<ComboBoxProps, 'values'> &
  AutoCompleteComboBox) => {
  // i18n
  const translateLanguage = useKoodistoLanguagesTranslation();

  // Helpers
  const filterSelectedLang = (
    excludedLanguage: string | undefined,
    valuesArray: Array<ComboBoxOption>,
    primaryLanguages: string[]
  ): Array<ComboBoxOption> => {
    const valuesArrayWithoutSelectedLang = valuesArray.filter(
      ({ value }) => value !== excludedLanguage
    );
    if (excludedLanguage && !primaryLanguages.includes(excludedLanguage)) {
      return valuesArrayWithoutSelectedLang.filter(({ value }) =>
        primaryLanguages.includes(value)
      );
    }

    return valuesArrayWithoutSelectedLang;
  };

  const values = languages.map((l) =>
    languageToComboBoxOption(translateLanguage, l)
  );

  const filteredValuesArray = filterSelectedLang(
    excludedLanguage,
    values,
    primaryLanguages
  );
  const optionValuesToShow = sortOptionsByLabels(filteredValuesArray);
  // Sort option value pairs into order set in primaryOptions parameter
  const primaryValues = optionValuesToShow
    .filter(({ value }) => {
      return primaryLanguages.indexOf(value) >= 0;
    })
    .sort((a, b) => {
      return (
        primaryLanguages.indexOf(a.value) - primaryLanguages.indexOf(b.value)
      );
    });

  // Merge sorted primaryOptions and sorted values
  const valuesToShow = [
    ...primaryValues,
    ...optionValuesToShow.filter((value) => !primaryValues.includes(value)),
  ];

  return <ComboBox {...rest} values={valuesToShow} />;
};
