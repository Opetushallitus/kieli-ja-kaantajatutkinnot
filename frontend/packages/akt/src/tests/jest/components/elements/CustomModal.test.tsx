import { render } from '@testing-library/react';

import { CustomModal } from 'components/elements/CustomModal';

// TODO The snapshots created look sensible, but rendering directly into document.body
// is discouraged. Figure out a better way to render the modals.

describe('CustomModal', () => {
  it('should render correctly when not open', () => {
    const { container } = render(
      <CustomModal
        open={false}
        onCloseModal={() => {
          return;
        }}
      >
        <div>Sisältöä tarvitaan</div>
      </CustomModal>,
      { container: document.body }
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when it is opened', () => {
    const { container } = render(
      <CustomModal
        open={true}
        onCloseModal={() => {
          return;
        }}
        ariaLabelledBy="testAriaLabel"
        screenReaderTitle="This is a screen reader title"
        modalTitle="Title for the modal"
      >
        <div>Lisää tekstiä</div>
      </CustomModal>,
      { container: document.body }
    );
    expect(container).toMatchSnapshot();
  });
});
