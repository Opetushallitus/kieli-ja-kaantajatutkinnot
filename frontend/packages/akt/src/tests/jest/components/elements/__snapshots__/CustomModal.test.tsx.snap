// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomModal should render correctly when it is opened 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-labelledby="testAriaLabel"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="custom-modal"
      tabindex="-1"
    >
      <div
        class="rows gapped"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
          id="testAriaLabel"
        >
          Title for the modal
        </h2>
        <div>
          Lisää tekstiä
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`CustomModal should render correctly when not open 1`] = `
<body>
  <div />
</body>
`;
